subsys/nrf_rpc/nrf_rpc_uart.c

```diff
@@ -127,19 +128,27 @@ static uint16_t tx_flip(struct nrf_rpc_uart *uart_tr, uint16_t crc_val)
 
 static bool rx_flip_check(struct nrf_rpc_uart *uart_tr, uint16_t crc_val)
 {
+	uint16_t last_rx_crc;
+
 	if (!IS_ENABLED(CONFIG_NRF_RPC_UART_RELIABLE)) {
 		return false;
 	}
 
+	last_rx_crc = uart_tr->flips.last_rx_crc;
+	uart_tr->flips.last_rx_crc = crc_val;
+
+	if (uart_tr->flips.rx_flip == FLIP_ANY ||
```

Comment:
I think this can be simplified. This checks independently the most significant bit and 15 least significant bits of the incoming CRC field while in fact it wants to just returns true iff `uart_tr.last_rx_crc_valid && last_rx_cr_valid == crc_val`, right?

Reply:
If the first frame has crc == 0 then it will be lost. I guess FLIP_ANY still plays its role to not lose the first frame.
In general, your assumption works for 99,999... % cases. :)

Reply:
Yes, that's why I added `uart_tr.last_rx_crc_valid` which would be equivalent of `!FLIP_ANY`. The thing is that you don't need to save the next expected flip bit if you already save the entire last CRC.

Reply:
I still do not understand it. `uart_tr.last_rx_crc_valid` should have some initial value, If first received frame has crc equal to initial value of `uart_tr.last_rx_crc_valid` it will be lost. `FLIP_ANY` is just property that say frame is first or not first.

Reply:
ok, you don't need `rx_flip` at all. You can just maintain:
bool last_rx_crc_valid : 1;
uint16_t last_rx_crc;

`last_rx_crc_valid` is initialized to `false`.
Then on each received frame you do:
```
bool duplicated;

duplicated = last_rx_crc_valid && (last_rx_crc == crc_val);
last_rx_crc_valid = true;
last_rx_crc = crc_val;

return duplicated;
```

Reply:
Nop, this will not work either. CRC16 is just hash that will not be unique for every byte sequence.
It is possible to have a situation when two different frames will have the same crc. If these frame came sequentially, then the second one will be filtered as duplicated.
I'm quite sure that idea of flipping is correct and should be used here.

Reply:
Sure but `crc_val` already contains the flipped bit. It's the sender that ensures that `crc_val` is unique for two subsequent frames sent. I don't propose to change the design, only simplify the implementation :).

Reply:
Ok, probably I got what you meant. Please take a look at the changes.

Reply:
Thanks, looks great.